### DISCUSSION TITLE
Welcome to TurnVoice Discussions!

### DISCUSSION THREAD

**Discussion by KoljaB** [KoljaB/TurnVoice#1] [Announcements]
## Welcome!
This GitHub Discussion space is designated for users and contributors of the project.

## Purpose:
Discuss applications and use-cases.
Collaborate on feature development and improvements.
Offer and seek technical support.
Discuss industry best practices related to low-latency voice technology.

## Participation Guidelines:
Ask questions related to the project's application and troubleshooting.
Share ideas for new features or improvements.
Upvote helpful discussions and answers.
Maintain a respectful and open-minded tone.

**Comment by dakinevich** [comment]
cmd always prints me Welcome to TurnVoice! But all tests runs okay. I cant understand how work with it.

**Reply by gellote** [reply]
the same happens to me

**Reply by KoljaB** [reply]
Ok, let's see. What exactly did you type in? Can you try the workflow described [here](https://github.com/KoljaB/TurnVoice?tab=readme-ov-file#preparation)? What is the output if you try to prepare a script with "turnvoice https://www.youtube.com/watch?v=cOg4J1PxU0c --prepare"? Does it only print "Welcome to TurnVoice!" and that's all or is there any kind of additional output you could post here (any kind of error message)? Does it freeze or return back to CLI?

**Reply by dakinevich** [reply]
the first problem I found I didn't have installed was the spleeter library. Further in cut.py on line 190, he did not replace the path to ffmpeg with the path to the exe file (although he added it to the Path and everything works). The problem is that the command line does not display an error message at all, so I created a python file and debug it using debug.  So far, I've stopped at hanging with the message switching “engine to coqui, voice male.wav” and line return CoquiEngine(language=self.language)

**Reply by dakinevich** [reply]
it print Welcome to TurnVoice! and back to CLI

**Reply by KoljaB** [reply]
May have something to do with the recent [coqui shutdown](https://coqui.ai/). Sounds like something does not get installed properly anymore. I need to test this on a clean environment next days.

**Reply by dakinevich** [reply]
what engine use instead of coqui?

**Reply by KoljaB** [reply]
OpenAI TTS, Azure, Elevenlabs or system voices.  Can be set with --engine parameter, it's described in the Readme.MD file

**Reply by dakinevich** [reply]
Changing the engine to system and fixing the errors that I listed helped me, now everything seems to work even on 6 gb vram

**Reply by KoljaB** [reply]
Nice, that's at least somehow good news. Sorry for the the missing coqui engine, I tried to reproduce the coqui problem, but worked here on a fresh env. There must be some thing I'm missing here, can't work it out right now. If possible, could you try out the [coqui test file](https://github.com/KoljaB/RealtimeTTS/blob/master/tests/coqui_test.py) from the RealtimeTTS lib used for synthesis? It's quite minimal code with some extended logging which could help track down the problem with coqui engine.

**Reply by dakinevich** [reply]
I'm interested in porting to Android, can you advise me on this matter?

**Reply by KoljaB** [reply]
I'm no Android expert but I think this is nearly impossible. Besides the turnvoice core there are some quite heavyweight libs used in the tech stack which you'd need to port to Android too.

From the requirements:

moviepy==1.0.3
yt-dlp==2023.11.16
faster_whisper==0.10.0
stable-ts==2.13.7
realtimetts==0.3.41	
textdistance==4.6.0
soundfile==0.12.1
pyannote.audio==3.1.1
deep-translator==1.11.4
instructor==0.4.2
demucs==4.0.1

Basically bringing only one of them to android would be quite a task.

**Reply by dakinevich** [reply]
Okay thanks for the answer

**Comment by Begia** [comment]
Hello! I'm also interested in trying out this program. However, after installation I'm always only getting the "Welcome to Turnvoice!" message when I'm trying out any examples. I've installed the program through pip (as instructed), and I get the help message if I type "turnvoice -h". Nothing else seems to work. For example, the line "turnvoice "https://www.youtube.com/watch?v=cOg4J1PxU0c" --prepare" only outputs "Welcome to Turnvoice!" and stops. I also tried installing spleeter, and running with the parameter "--engine system". I'm running this in Garuda linux with the fish shell, and using micromamba.

**Reply by KoljaB** [reply]
Can't support this repo currently, too much to do. Also I'm no linux expert.

Try: 
- open https://github.com/KoljaB/TurnVoice/blob/main/turnvoice/core/prepare.py
- put print statements there and look at which parts of the main program gets passed to get to the failing part.

**Reply by Begia** [reply]
Hi, Sorry I don't follow. Should I edit the prepare.py in my own installation? Where is it possible to put print statements in the code?

I don't understand the framework exactly, so I'm not sure how the program runs this prepare.py script. I'm not very well-versed in python applications myself. Thanks for your help!

**Reply by KoljaB** [reply]
Can't give much support currently, I can hardly type (health issues). Without some experience in python I think it will be hard to get it run, maybe even with the recommended environment (some dependencies often still vary)

**Reply by Begia** [reply]
Okay, I hope you get better soon! I've done a lot of things with python, just haven't really made stand-alone programs (with inputs and outputs as parameters)), so I don't understand the architechture. I think I've pinpointed the source of the issue to the spleeter package and seems like it didn't install even when I thought it did. Now I'm having issues with tons of conflicting dependencies, which seem to prevent installation of spleeter in the same conda environment. Trying to work through this somehow :)

**Reply by KoljaB** [reply]
Thx. Might be that spleeter is not even needed any more. I think I switched to demux from meta, maybe worth a try. 
Python code for this is in cut.py

**Reply by Begia** [reply]
Yes, seems like cut.py has only the demux mentioned. However, I'm still not able to run and I seem to have all the dependencies installed outside of spleeter. The latest version of spleeter requires python < 3.11, so I'm wondering how is it possible to run this with the recommended and tested python 3.11.4 (which I now have in my conda environment). I agree that the dependecies of turnvoice are a real jungle! I think it might be possible to somehow fork the code and call the spleeter functions from another environment or even a running docker container. However, I haven't found where these commands are in your code, since I'm not familiar with the program architecture. 

If you have some kind of a environment yml file or such of all the packages and their versions, that could be used to recreate your exact testing environment in conda, though.

Maybe I just need to give up at this point and try something else...

**Reply by KoljaB** [reply]
I think spleeter needs to be installed in a venv.

**Reply by KoljaB** [reply]
Part of the problem is python. I can set fixed versions of dependent libs like in Turnvoice but if these libs use other libs and don't fix the versions too then dependency hell starts.

**Reply by KoljaB** [reply]
Here is a copy/paste of the full requirements.txt with all libs from a working venv fwiw:

```bash
absl-py==2.0.0
aiohttp==3.9.1
aiosignal==1.3.1
alembic==1.13.1
annotated-types==0.6.0
antlr4-python3-runtime==4.9.3
anyascii==0.3.2
anyio==4.2.0
asteroid-filterbanks==0.4.0
asttokens==2.4.1
attrs==23.1.0
audioread==3.0.1
av==10.0.0
azure-cognitiveservices-speech==1.36.0
Babel==2.14.0
bangla==0.0.2
beautifulsoup4==4.12.2
blinker==1.7.0
blis==0.7.11
bnnumerizer==0.0.2
bnunicodenormalizer==0.1.6
Brotli==1.1.0
cachetools==5.3.2
catalogue==2.0.10
certifi==2023.11.17
cffi==1.16.0
charset-normalizer==3.3.2
click==8.1.7
cloudpathlib==0.16.0
cloudpickle==3.0.0
colorama==0.4.6
coloredlogs==15.0.1
colorlog==6.8.0
comtypes==1.2.0
confection==0.1.4
contourpy==1.2.0
coqpit==0.0.17
ctranslate2==3.23.0
cycler==0.12.1
cymem==2.0.8
Cython==3.0.7
dateparser==1.1.8
decorator==4.4.2
deep-translator==1.11.4
demucs==4.0.1
distro==1.8.0
docopt==0.6.2
docstring-parser==0.15
dora-search==0.1.12
einops==0.7.0
elevenlabs==0.2.27
emoji==2.8.0
encodec==0.1.1
executing==2.0.1
faster-whisper==0.10.0
ffmpeg-python==0.2.0
filelock==3.13.1
Flask==3.0.0
flatbuffers==23.5.26
fonttools==4.47.0
frozenlist==1.4.1
fsspec==2023.12.2
future==0.18.3
g2pkk==0.1.2
google-auth==2.25.2
google-auth-oauthlib==1.2.0
greenlet==3.0.2
grpcio==1.60.0
gruut==2.2.3
gruut-ipa==0.13.0
gruut-lang-de==2.0.0
gruut-lang-en==2.0.0
gruut-lang-es==2.0.0
gruut-lang-fr==2.0.2
h11==0.14.0
hangul-romanize==0.1.0
httpcore==1.0.2
httpx==0.26.0
huggingface-hub==0.20.1
humanfriendly==10.0
HyperPyYAML==1.2.2
idna==3.6
imageio==2.33.1
imageio-ffmpeg==0.4.9
inflect==7.0.0
instructor==0.4.2
ipython==8.18.1
itsdangerous==2.1.2
jamo==0.4.1
jedi==0.19.1
jieba==0.42.1
Jinja2==3.1.2
joblib==1.3.2
jsonlines==1.2.0
julius==0.2.7
kiwisolver==1.4.5
lameenc==1.7.0
langcodes==3.3.0
lazy_loader==0.3
librosa==0.10.1
lightning==2.1.2
lightning-utilities==0.10.0
llvmlite==0.41.1
Mako==1.3.0
Markdown==3.5.1
markdown-it-py==3.0.0
MarkupSafe==2.1.3
matplotlib==3.8.2
matplotlib-inline==0.1.6
mdurl==0.1.2
more-itertools==10.1.0
moviepy==1.0.3
mpmath==1.3.0
msgpack==1.0.7
multidict==6.0.4
murmurhash==1.0.10
mutagen==1.47.0
networkx==2.8.8
nltk==3.8.1
num2words==0.5.13
numba==0.58.1
numpy==1.26.2
oauthlib==3.2.2
omegaconf==2.3.0
onnxruntime==1.16.3
openai==1.13.3
openai-whisper==20231117
openunmix==1.2.1
optuna==3.5.0
packaging==23.2
pandas==1.5.3
parso==0.8.3
Pillow==10.1.0
platformdirs==4.1.0
pooch==1.8.0
preshed==3.0.9
primePy==1.3
proglog==0.1.10
prompt-toolkit==3.0.43
protobuf==4.23.4
psutil==5.9.7
pure-eval==0.2.2
pyannote.audio==3.1.1
pyannote.core==5.0.0
pyannote.database==5.0.1
pyannote.metrics==3.2.1
pyannote.pipeline==3.0.1
pyasn1==0.5.1
pyasn1-modules==0.3.0
PyAudio==0.2.14
pycparser==2.21
pycryptodomex==3.19.0
pydantic==2.5.2
pydantic_core==2.14.5
pydub==0.25.1
Pygments==2.17.2
pynndescent==0.5.11
pyparsing==3.1.1
pypinyin==0.50.0
pypiwin32==223
pyreadline3==3.4.1
pysbd==0.3.4
python-crfsuite==0.9.10
python-dateutil==2.8.2
pytorch-lightning==2.1.2
pytorch-metric-learning==2.4.1
pyttsx3==2.90
pytz==2023.3.post1
pywin32==306
PyYAML==6.0.1
RealTimeTTS==0.3.43
regex==2023.10.3
requests==2.31.0
requests-oauthlib==1.3.1
retrying==1.3.4
rich==13.7.0
rsa==4.9
ruamel.yaml==0.18.5
ruamel.yaml.clib==0.2.8
safetensors==0.4.1
scikit-learn==1.3.2
scipy==1.11.4
semver==3.0.2
sentencepiece==0.1.99
shellingham==1.5.4
six==1.16.0
smart-open==6.4.0
sniffio==1.3.0
sortedcontainers==2.4.0
soundfile==0.12.1
soupsieve==2.5
soxr==0.3.7
spacy==3.7.2
spacy-legacy==3.0.12
spacy-loggers==1.0.5
speechbrain==0.5.16
SQLAlchemy==2.0.23
srsly==2.4.8
stable-ts==2.13.7
stack-data==0.6.3
stanza==1.6.1
stream2sentence==0.2.3
submitit==1.5.1
SudachiDict-core==20230927
SudachiPy==0.6.8
sympy==1.12
tabulate==0.9.0
tensorboard==2.15.1
tensorboard-data-server==0.7.2
tensorboardX==2.6.2.2
textdistance==4.6.0
thinc==8.2.2
threadpoolctl==3.2.0
tiktoken==0.5.2
tokenizers==0.15.0
torch==2.1.2+cu118
torch-audiomentations==0.11.0
torch-pitch-shift==1.2.4
torchaudio==2.1.2+cu118
torchmetrics==1.2.1
tqdm==4.66.2
trainer==0.0.36
traitlets==5.14.0
transformers==4.36.2
treetable==0.2.5
TTS==0.22.0
# Editable Git install with no remote (TurnVoice==0.0.65)
-e d:\projekte\turnvoice
typer==0.9.0
typing_extensions==4.9.0
tzdata==2023.3
tzlocal==5.2
umap-learn==0.5.5
Unidecode==1.3.7
urllib3==2.1.0
wasabi==1.1.2
wcwidth==0.2.12
weasel==0.3.4
websockets==12.0
Werkzeug==3.0.1
yarl==1.9.4
yt-dlp==2023.11.16

```

**Reply by Begia** [reply]
Thanks again for your trouble! I tried making a new virtual environment directly with this package list, but it appears unsolvable. 

I guess building a docker container for turnvoice would be the easiest solution to make it directly available in a working configuration. For now, I think I need to give up on trying to install it.

**Comment by Lucidology** [comment]
I have this problem as well, "Welcome to TurnVoice!" and stops. GPT suggested this as a possible solution, maybe the author can use poetry can make a poetry.lock that gets all the deeper dependencies:

Poetry: Manages dependencies and can resolve the dependency tree to prevent conflicts before installing packages. It also locks dependencies to specific versions via poetry.lock to ensure reproducible builds.

**Reply by Lucidology** [reply]
Here are the commands I've tried. I noticed if I gave it a local file it uses the disk a lot before quitting. So maybe adding a command line option for "verbose logging" to trace what it's doing would help diagnose problems in the field.

```
R:\-\r browser downloads\v>turnvoice -i "ESP Programmer for ESP32 and ESP-12E -ESP8266- Microcontroller.mp4"  -v arthur.wav -o cooking_with_arthur.mp4
Welcome to TurnVoice!

R:\-\r browser downloads\v>turnvoice -i AmC9SmCBUj4 -v arthur.wav -o cooking_with_arthur.mp4
Welcome to TurnVoice!

R:\-\r browser downloads\v>turnvoice -i AmC9SmCBUj4 -v arthur.wav -o cooking_with_arthur.mp4 -e system
Welcome to TurnVoice!

R:\-\r browser downloads\v>turnvoice -i AmC9SmCBUj4 -v David -o cooking_with_arthur.mp4 -e system
Welcome to TurnVoice!
```
